### PR TITLE
Improve Field class

### PR DESCRIPTION
**Pull request by Florisheinen1**
### A clean new Field class
To put it bluntly: the current Field class is a mess. It's big(500 lines), has many practically unused variables, and much unnecessary wrapping going on. (And what is [that map](https://github.com/RoboTeamTwente/roboteam/blob/a64f0db278e06de9661181b05681949e7084052f/roboteam_ai/include/roboteam_ai/world/Field.h#L69) doing!?) I think simplifying this whole class leads to higher efficiency in both processing and our development(less confusing, more programming brrrrrr, less bugs). That is why I made this PR.

#### Changes
The goal of this PR was to abstract away all billion individual lines and points into just a few rectangles. This prevents having to give each existing point and line a specific name (which gets quite confusing). All these rectangles already have the [built in `contains`](https://github.com/RoboTeamTwente/roboteam/blob/3994422aa3dbcfc04c5b82f026287f852260bd75/roboteam_utils/include/roboteam_utils/FastRectangle.hpp#L25) function etc. which removes the need to have to [manually write one](https://github.com/RoboTeamTwente/roboteam/blob/a64f0db278e06de9661181b05681949e7084052f/roboteam_ai/include/roboteam_ai/world/FieldComputations.h#L48) for each area in the field.

With this comes the switch from using length and width to using width and height. Reasons for this transition: 
1. Sometimes with weird field setups, the long side is on the y-axis instead of the x-axis. This renders the name length for x-axis and width for y-axis incorrect (or, at least confusing).
2. Rectangles by default have a width/height approach, as that is the default. Having to transition everytime in your code between the fields' length/width to its Rectangle's width/height leads to confusion([here someone mixed the wrong and right width](https://github.com/RoboTeamTwente/roboteam/blob/a64f0db278e06de9661181b05681949e7084052f/roboteam_ai/src/utilities/Dealer.cpp#L271)). I could have made the Rectangles also use the length/width approach, but I want the Rectangle class to be usable in different situations where the length/width approach does not make any sense (think of GUI's, or the goal area). In conclusion, a consistent approach is necessary, and the width/height one is preferrable.

#### For the reviewer
Because this changes a lot of code, I tried to make each commit very clear. Some small commits change big things, while big commits only refactor names of functions and variables etc. I would suggest going through each commit one by one to check the changes made instead of looking at the "all files changed", as that is waaay too much to review. If you are too lazy to go through it yourself, I am always happy to go to the office and explain everything in person :D

#### Extra examples of why this new Field is nice
1. Everything in the old field was wrapped in optional. First of all, people tend to forget to check if the value is actually set. Second of all, if you know the left goal is defined, you can bet money on the fact the right goal is defined: Either the whole field is present or absent. (+performance gain)
2. New Field is now in utils. This means roboteam_observer can use it too :)
4. Width/height is used more often, so [people can forget width is for the y axis](https://github.com/RoboTeamTwente/roboteam/blob/a64f0db278e06de9661181b05681949e7084052f/roboteam_ai/src/utilities/Dealer.cpp#L271) (example where width and length is accidentally swapped)
5. [Converting from length/width to width/height](https://github.com/RoboTeamTwente/roboteam/blob/a64f0db278e06de9661181b05681949e7084052f/roboteam_ai/src/interface/widgets/widget.cpp#L169) is now luckily no longer necessary.
6. This new Field class is easier to test. No need to write tests for if points and lines are set correctly, or if our manual `pointInOurDefenseArea` is correct, or comparing different Fields
7. I know I forgot a few extra benefits, but I forgot them as it is already quite late

###### Code Quality
- [ ] Is the code is understandable and easy to read
- [ ] Changes to the code comply with set clang-format rules
- [ ] No use of manual memory control (e.g new/malloc/colloc etc)
- [ ] Are (only) smart pointers used?

###### Testing
- [ ] All tests are passing.
- [ ] I _added new / changed existing_ tests to reflect code changes (state why not otherwise!)
- [ ] I tested my changes manually (Describe how, to what extent etc.)

###### Commit Messages
- [ ] Commit message is saying what has been changed, **why** it was changed? Remember other developers might not know
  what the problem you are fixing was. Note also negative _decision_ (e.g., why did you not do particular thing)
  **TLDR: Commit message are comprehensive**
- [ ] Commit messages follows the rules of https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Review by rolfvdhulst — CHANGES_REQUESTED** [review]
Generally looks good, I agree with most changes/ the direction this is going.
I would suggest two things: 
1. Remove the Grids from the Field class. They are not part of the 'geometric information' which is the field, and are only there for the plays. Additionally, since these can become quite large, they make the whole application slower as the field is (unfortunately) copied around quite a bit.

2. Keep some of the aliasses for ourGoalCenter() etc. The extra mental overhead of leftGoalArea.rightline().center() starts to add up and makes the code less readible, in my opinion.

The last thing I noticed is that the Field and the world are being copied and passed around manually A LOT in a lot of different ways (if I counted correctly, each play has access to potentially 4 distinct fields) which it really should not be as it is basically constant. However, this is not a problem of your PR to solve but rather the reality of the current system, and should be solved in a new PR which addresses these structural information flow issues.

File: roboteam_ai/include/roboteam_ai/stp/Play.hpp [hunk 1]
```diff
@@ -162,7 +162,7 @@ class Play {
     /**
      * The Field
      */
-    rtt::world::Field field;
+    rtt::Field field;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/include/roboteam_ai/stp/Play.hpp:165]: Agree on the change and the fact that it should be in a different PR

File: roboteam_utils/include/roboteam_utils/Field.hpp [hunk 2]
```diff
@@ -20,13 +20,24 @@ typedef struct Field {
...
+    // TODO: Put these grids into a single grid of grids
+    Grid topLeftGrid;
+    Grid topMidGrid;
+    Grid topRightGrid;
+    Grid middleLeftGrid;
+    Grid middleMidGrid;
+    Grid middleRightGrid;
+    Grid bottomLeftGrid;
+    Grid bottomMidGrid;
+    Grid bottomRightGrid;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_utils/include/roboteam_utils/Field.hpp:40]: I fully agree that the Grids do not belong in the Field class. However, the purpose of this PR was not to change the old Field, but merely to clean it up. That unfortunately also means copying over the grids. I did look into removing them, but that would have required waaaay more code to be altered throughout the codebase (so, a different PR it is)

File: roboteam_utils/src/utils/Field.cpp [hunk 3]
```diff
@@ -13,7 +13,17 @@ bool Field::operator==(const Field &other) const {
...
+        && this->rightGoalArea == other.rightGoalArea
+        && this->bottomLeftGrid == other.bottomLeftGrid
+        && this->bottomMidGrid == other.bottomMidGrid
+        && this->bottomRightGrid == other.bottomRightGrid
+        && this->middleLeftGrid == other.middleLeftGrid
+        && this->middleMidGrid == other.middleMidGrid
+        && this->middleRightGrid == other.middleRightGrid
+        && this->topLeftGrid == other.topLeftGrid
+        && this->topMidGrid == other.topMidGrid
+        && this->topRightGrid == other.topRightGrid;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_utils/src/utils/Field.cpp:25]: Yes, agree on this too, but for tests it is convenient to check if the grids also match when comparing fields. On top of that, it might be misleading to use the operator== function for comparing field objects without checking the grids. But this will all be solved when the grids are no longer part of the Field class. TODO!

File: roboteam_utils/src/utils/Field.cpp [hunk 4]
```diff
@@ -50,15 +60,41 @@ Field Field::createField(double fieldWidth, double fieldHeight, double defenseWi
...
-        .rightDefenseArea  = FieldRectangle(g, h),
-        .leftGoalArea      = FieldRectangle(i, j),
-        .rightGoalArea     = FieldRectangle(k, l),
+        .playArea          = FastRectangle(a, d),
+        .leftPlayArea      = FastRectangle(a, b),
+        .rightPlayArea     = FastRectangle(c, d),
+        .leftDefenseArea   = FastRectangle(e, f),
+        .rightDefenseArea  = FastRectangle(g, h),
+        .leftGoalArea      = FastRectangle(i, j),
+        .rightGoalArea     = FastRectangle(k, l),
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_utils/src/utils/Field.cpp:69]: The variable names ('a', 'b' etc. ) _are_ in fact made by me. And in my opinion, simply using these letters as variable names is more clear than using very long specific names. (Keep in mind of course the big comment block explaining them above this piece of code).
I do agree that using explaining variable names is usually good, but this is one of those rare exceptions:
- Names are contained within the function. The only situation where you encounter these variables is when you are looking into how the function works, and when you do that, you will realize why and how they work the way they work.
- Using long variable names might clutter your view of the code, while moving your eyes a few lines up gives you a visual explanation of what the short names mean.

In conclusion, I think both ways of naming variables are applicable in this situation, so I leave it up to the current team to make the decision, and will not change it myself.

File: roboteam_ai/include/roboteam_ai/stp/StpInfo.h [hunk 5]
```diff
@@ -110,7 +110,7 @@ struct StpInfo {
     /**
      * Field
      */
-    std::optional<world::Field> field;
+    std::optional<Field> field;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/include/roboteam_ai/stp/StpInfo.h:113]: Yes

File: roboteam_ai/include/roboteam_ai/utilities/Dealer.h [hunk 6]
```diff
@@ -89,7 +89,7 @@ class Dealer {
 
    private:
     v::WorldDataView world;
-    rtt::world::Field *field;
+    rtt::Field *field;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/include/roboteam_ai/utilities/Dealer.h:92]: Yes

File: roboteam_ai/include/roboteam_ai/world/World.hpp [hunk 7]
```diff
@@ -216,7 +216,7 @@ class World {
      */
     std::optional<WorldData> currentWorld = std::nullopt;
 
-    std::optional<world::Field> currentField;
+    std::optional<Field> currentField;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/include/roboteam_ai/world/World.hpp:219]: Yes

File: roboteam_ai/src/interface/widgets/PlaysWidget.cpp [hunk 8]
```diff
@@ -9,7 +9,7 @@
 namespace rtt::ai::interface {
 inline QString formatPlay(stp::Play* play) {
     std::optional<rtt::world::view::WorldDataView> world;
-    std::optional<rtt::world::Field> field;
+    std::optional<rtt::Field> field;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/interface/widgets/PlaysWidget.cpp:12]: Yes

File: roboteam_ai/src/interface/widgets/RobotsWidget.cpp [hunk 9]
```diff
@@ -26,7 +26,7 @@ RobotsWidget::RobotsWidget(QWidget *parent) : QWidget(parent) {
 }
 
 void RobotsWidget::updateContents(Visualizer *visualizer, rtt::world::view::WorldDataView world) {
-    std::optional<rtt::world::Field> field;
+    std::optional<rtt::Field> field;
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/interface/widgets/RobotsWidget.cpp:29]: Yes

File: roboteam_utils/test/utils/GridTest.cpp [hunk 10]
```diff
@@ -37,16 +37,16 @@ TEST(Grid, getters) {
...
-    EXPECT_DOUBLE_EQ(grid.getStepSizeX(), 4 / 5.0);
-    EXPECT_DOUBLE_EQ(grid.getStepSizeY(), 3 / 6.0);
+    EXPECT_DOUBLE_EQ(grid.getStepSizeX(), 3 / 5.0);
+    EXPECT_DOUBLE_EQ(grid.getStepSizeY(), 4 / 6.0);
 }
 
 // Test whether the points are equally spaced in the region (average should be in the center of the grid)
 TEST(Grid, spreadTest) {
-    Grid grid = Grid(-3, -2, 4, 6, 5, 3);
+    Grid grid = Grid(-3, -2, 6, 4, 5, 3);
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_utils/test/utils/GridTest.cpp:49]: The Grid class used to have width(y) and length(x) as parameter, but it got swapped with width(x) and height(y), resulting in a change of order in parameter, hence the swapping of the 4 and the 6.

File: roboteam_ai/src/world/World.cpp [hunk 11]
```diff
@@ -66,12 +66,45 @@ void World::updateWorld(proto::World &protoWorld) {
...
+}
+
 void World::updateField(proto::SSL_GeometryFieldSize &protoField) {
-    world::Field field(protoField);
-    this->currentField = field;
+
+    // TODO: Move this conversion from the scary proto to the safe c++ rtt::Field type in observer or something.
+    // Observers' whole function is to filter and parse anyways, might as well check which proto fields are set and which arent...
+    // Also, the proto does not require penalty_area_depth to be set *UNLIKE GOAL DEPTH AND WIDTH*,
+    // so we need to derive specifically these lines from a list of named lines... Kill me plz
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/world/World.cpp:79]: Will do!

File: roboteam_ai/src/stp/plays/referee_specific/KickOffThemPrepare.cpp [hunk 12]
```diff
@@ -26,32 +26,32 @@ KickOffThemPrepare::KickOffThemPrepare() : Play() {
...
-    stpInfos["formation_5"].setPositionToMoveTo(Vector2(defense_line_x, width / 5));
-    stpInfos["formation_6"].setPositionToMoveTo(Vector2(defense_line_x, -width / 5));
-    stpInfos["formation_7"].setPositionToMoveTo(Vector2(-length / 4, width / 3));
-    stpInfos["formation_8"].setPositionToMoveTo(Vector2(-length / 4, -width / 3));
-    stpInfos["formation_9"].setPositionToMoveTo(Vector2(-length * 3.0 / 16.0, 0.0));
+    stpInfos["formation_5"].setPositionToMoveTo(Vector2(defense_line_x, height / 5));
+    stpInfos["formation_6"].setPositionToMoveTo(Vector2(defense_line_x, -height / 5));
+    stpInfos["formation_7"].setPositionToMoveTo(Vector2(-width / 4, height / 3));
+    stpInfos["formation_8"].setPositionToMoveTo(Vector2(-width / 4, -height / 3));
+    stpInfos["formation_9"].setPositionToMoveTo(Vector2(-width * 3.0 / 16.0, 0.0));
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/stp/plays/referee_specific/KickOffThemPrepare.cpp:54]: Yes, I am sure this is correct, as it is not the height that changed into length, but width to height and length to width

File: roboteam_ai/src/world/World.cpp [hunk 13]
```diff
@@ -66,12 +66,45 @@ void World::updateWorld(proto::World &protoWorld) {
     setWorld(data);
 }
 
+// Converts millimeters to meters
+double mmToM(double mm) {
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/world/World.cpp:70]: Will do! (I think this will be a temporary function anyways, as this is only used in the conversion between proto and c++ Fields, which hopefully moves to Observer)

File: roboteam_ai/src/stp/plays/referee_specific/PenaltyUsPrepare.cpp [hunk 14]
```diff
@@ -23,15 +23,15 @@ PenaltyUsPrepare::PenaltyUsPrepare() : Play() {
...
+uint8_t PenaltyUsPrepare::score(const rtt::Field& field) noexcept {
     /// List of all factors that combined results in an evaluation how good the play is.
     scoring = {{PlayEvaluator::getGlobalEvaluation(eval::PenaltyUsPrepareGameState, world), 1.0}};
     return (lastScore = PlayEvaluator::calculateScore(scoring)).value();  // DONT TOUCH.
 }
 
 void PenaltyUsPrepare::calculateInfoForRoles() noexcept {
     // We need at least a keeper, and a kicker positioned behind the ball
-    stpInfos["keeper"].setPositionToMoveTo(Vector2(field.getOurGoalCenter()));
+    stpInfos["keeper"].setPositionToMoveTo(Vector2(field.leftGoalArea.rightLine().center()));
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/stp/plays/referee_specific/PenaltyUsPrepare.cpp:34]: Hmmm, I guess you might have a point. In my opinion, the `leftGoalArea.rightLine().center()` is more clear than the `getOurGoalCenter()`, because to me it was unclear whether 'center' meant the center of the goal area or the center of the goal line, but I do see how that might differ per person. I will again leave that up to the current team.

File: roboteam_utils/CMakeLists.txt [hunk 15]
```diff
@@ -28,7 +28,7 @@ set(ROBOTEAM_UTILS_SRC
         src/utils/FileLogger.cpp
         src/utils/Teams.cpp
         src/utils/AIData.cpp
-        src/utils/FieldRectangle.cpp
+        src/utils/FastRectangle.cpp
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_utils/CMakeLists.txt:31]: That would also mean all the other header files should be included in the CMakeLists file. I am not sure if it will in fact change anything at all, but I guess it might be nice? In any case, I consider that not part of this PR.

File: roboteam_ai/src/control/positionControl/pathPlanning/NumTreesPlanning.cpp [hunk 16]
```diff
@@ -71,7 +71,7 @@ std::vector<Vector2> NumTreesPlanning::computePath(const Vector2 &robotPosition,
 }
 
 std::vector<PathPointNode> NumTreesPlanning::branchPath(PathPointNode &parentPoint, const Vector2 &collisionPosition, Vector2 &destination) const {
-    // const rtt_world::Field &field, const Vector2 &point, bool isOurDefenceArea, double margin, double backMargin
+    // const rtt::Field &field, const Vector2 &point, bool isOurDefenceArea, double margin, double backMargin
```
[the remark this answers is not in the document]
Comment (Florisheinen1) [reply roboteam_ai/src/control/positionControl/pathPlanning/NumTreesPlanning.cpp:74]: Yes, that comment can be removed, which I will do!

**Comment by rolfvdhulst** [comment]
LGTM now, fine by me if removing the grid goes into another PR/Issue, this is at least a clear improvement already. I think someone else should take a look before it's merged too so they can approve it in that case.

File: roboteam_ai/src/stp/plays/referee_specific/PenaltyUsPrepare.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (JAndringa) [reply roboteam_ai/src/stp/plays/referee_specific/PenaltyUsPrepare.cpp:34]: I prefer the `leftGoalArea.rightLine().center()`, but as Floris said this personal preference. I also think keeping the old function reduces the readability of the Field class, which is the opposite of what we want.

**Review by JAndringa — APPROVED** [review]
In my opinion, width/length should be replaced with height/width everywhere for consistency, but perhaps it's better to have a separate PR for this. Other than that LGTM.

**Review by rolfvdhulst — APPROVED** [review]
